### PR TITLE
Switch loop marker shortcuts, add context menu

### PR DESCRIPTION
**Pull request by Spekular**
See #5505 and #6014

This PR now halfway-implements remappable loop marker controls, with limitations:
- Unmodified LMB and RMB are reserved (playhead, context menu)
- All combinations of Shift/Control and LMB/RMB can be assigned an action. For applicable actions (everything except selecting TCOs), this will be quantized
- Shift + Control + click is used for the unquantized version of Shift + Click actions

#### Future enhancements

- Bindings are currently assigned in code, but should be exposed via the UI. I'm not sure it's wise to add this to the current settings menu, so this might involve rewriting it or adding a new menu for keybindings.
- Allow the loop section to be shifted as a whole. Could be part of the drag loop mode.
- Populate the context menu, it currently only contains "select between loop markers".
  - Fit loop markers to selection
  - Zoom to loop/selection/song
  - Quantize loop marker length, start, end, start + end
  - Double/halve loop length

#### Proposed default shortcuts:
LMB: Move playhead
RMB: Open context menu

Ctrl + LMB: Group select
Ctrl + RMB: Mean select

Shift + LMB: Loop Start
Shift + RMB: Loop End

Holding both modifiers disables snapping. If both are held at the start of the action, the shift action is selected.

### CODE REVIEW

**Comment by Spekular** [comment]
I experimented a bit with rebinding support, but you'd either need to introduce severe restrictions or drastically rework the mouse event functions. In particular, the `m_action` system is an obstacle.

It's very easy to handle dragging the loop start, loop end, or playhead without it, just check mouse button and modifiers in the mouse move handler. Dragging with mean based logic should work without it if you always recalculate the closest point in the move event, but this could get hairy when the points approach each other. Selection needs state to remember where the selection started, and needs to emit selectionFinished when done.

In the end I decided it wasn't worth it, since there's not much difference between e.g. using ctrl or shift for an operation.

**Comment by superpaik** [comment]
It works just fine, as stated in the PR description.

Some thoughts about it.
- Would it be possible to have an option to move the whole loop marker? I mean, if you have a loop marker of, say, 8 beats, move/drag the whole loop marker (the use case would be move the loop marker from one drop/break/built to the next) As a key/mouse option I'd say LMB inside the loop marker and drag it.
-An as for the LMB option (move playhead), I wonder why the default option it isn't snap to the grid but fine snapping (it's something not introduced by this PR, though, it happens in 1.2.2 as well)? To me it's kind of odd behaviour because you normally want to move the playhead to the start of a beat. I'm not aware right now if there is an alternative to do that. Maybe it's out of the scope of this PR and needs a debate about it.

**Comment by Spekular** [comment]
I'd be very surprised to see the playhead quantized when moving. Even with mostly programmed stuff there are many elements that comes in slightly before grid lines (e.g. preshifted claps) that you can easily account for with an unquantized playhead. If you start working with recorded material things get even less perfectly aligned.

As for moving the loop, I agree that it'd be useful but I don't think unmodified LMB should do anything except playhead movement. If that's desireable then IMHO we should make a vertical division, so there's a playhead lane and a loop marker lane. Of course, that takes up more vertical space since you want the lanes to remain a comfortable height.

Maybe I need to give remapping some more thought.

**Comment by qnebra** [comment]
I think playhead position shouldn't be automatically quantized, especially when manually moved.

Maybe Ctrl+Shift+LMB inside loop to move entire loop section?

**Comment by superpaik** [comment]
I see your point regarding not quantizing the playhead, but I think that in any other use case the default behaviour is quantizing and you have an option for not quantizing it (like in the loop marker, midi/beat/sample track end, etc.) so thats why I said it, but it is ok to me, because I haven't notice it until I tested this PR and thought about how it should behave. As for context, Ableton quantize the playhead by default and with Alt+LMB you can move it freely. 

As for moving the whole loop the option Ctrl+Shift+LMB, as qnebra said, it's already taken for Loop Start (fine). The option Spekular propose would be perfect, but I guess it would require a major code change for just this functionality (unless is used for future functionalities like song markers). I don't know if ALT+LMB is an option because of cross platform issues, but another suggestion would be "doble-click" and move it until next click. Another option, which I think doesn't require a new lane, is that when inside (and only when inside) the loop marker and the cursor is at the top of the lane (above the playhead arrow) change the cursor to an double-arrow so you can move the loop with single LMB (see image). Perhaps with low resolution screens this may be an issue. 
imagen

Anyway, they where just thoughts that came to my mind while testing. Don't be too worried.

File: src/gui/TimeLineWidget.cpp
Comment (IanCaio) [inline src/gui/TimeLineWidget.cpp:338]: Second conditional can be one-lined too

**Review by IanCaio — COMMENTED** [review]
Still need to give a more thoughtful review and test it, but at first glance it looks good. There are some formatting changes I can later put on a review for a batch commit if you want. In this one I only addressed this conditional that can be one-lined to match the previous one.

**Comment by Spekular** [comment]
Ok, I came up with a binding system that I think strikes a nice balance between flexibility and amount of code. I've reworked the PR quite a bit and as such edited the main post. There's at least one bug that needs squashing before merge (assistance here would be awesome), and ideally the re-binding support would be full fledged as well.

File: src/gui/TimeLineWidget.cpp
Comment (IanCaio) [inline src/gui/TimeLineWidget.cpp:404]:
```suggestion
			tr("Hold <%1> to disable quantization.").arg(UI_CTRL_KEY),
			embed::getIconPixmap("hint"), 0);
```

Plus maybe having that hint box for Ctrl actions as well as suggested in the other comment

**Review by IanCaio — CHANGES_REQUESTED** [review]
Mostly style changes and some doubts/suggestions. The code looks good, behavior is nice given some time to get used to it.

I found the "DragLoop" action name a bit misleading at first. Until I downloaded the build and tested it I thought that action was supposed to drag both loop points simultaneously (thus dragging the loop). I'd maybe rename it to "DragSelectLoop", which sounds more like what it does.

File: src/gui/TimeLineWidget.cpp
Comment (Spekular) [reply src/gui/TimeLineWidget.cpp:367]: It does, but as far as I can tell moving the playhead still moves perfectly fine. Frankly I have no idea what `m_moveXOff` is supposed to do or in what edge case it ends up not being half the pixmap width. I've tried clicking within the playhead triangle, and clicking so that it's by the start of the song, and everything works.

File: src/gui/TimeLineWidget.cpp
Comment (Spekular) [reply src/gui/TimeLineWidget.cpp:383]: That wasn't the intended behavior, but yeah, that's how it turned out. That seems fine to me so I'll update the tooltip instead of trying to prevent it.

File: src/gui/TimeLineWidget.cpp
Comment (IanCaio) [inline src/gui/TimeLineWidget.cpp:426]:
```suggestion
		m_hint = nullptr;
```

File: src/gui/TimeLineWidget.cpp
Comment (IanCaio) [inline src/gui/TimeLineWidget.cpp:408]: For consistency, is there a `UI_SHIFT_KEY` constant that should be used here too? I'm not sure, just asking, so if there's not just ignore this comment.

File: src/gui/TimeLineWidget.cpp
Comment (IanCaio) [reply src/gui/TimeLineWidget.cpp:372]: I see, it's probably not noticeable because the width is very small. We should probably decide between both ways though. If it's being overridden without apparent issue, maybe we should just delete the lines that calculate the offset when it's smaller than the width of the marker.

**Review by IanCaio — COMMENTED** [review]
Sorry it took a while. On a second review nothing stood out besides those small comments!

File: src/gui/TimeLineWidget.cpp
Comment (Spekular) [reply src/gui/TimeLineWidget.cpp:408]: I'm pretty sure the constant exists from ctrl because macs call it command instead.

File: src/gui/TimeLineWidget.cpp
Comment (Spekular) [inline src/gui/TimeLineWidget.cpp:391]:
```suggestion
		m_loopPos[1] = m_loopPos[0] + (quant ? m_snapSize : 1);
```

**Review by IanCaio — APPROVED** [review]
I'd still rather we remove the irrelevant line (compiler might even be optimizing it out), but since it wasn't added by this PR particularly and everything else was fixed, I'm giving my approval!

File: src/gui/TimeLineWidget.cpp
Comment (Spekular) [inline src/gui/TimeLineWidget.cpp:334]:
```suggestion
	auto leftCtrlAction = SelectSongTCO;
```

**Comment by allejok96** [comment]
(To everyone but Spekular)

I've worked out a way to map RMB *click* to context menu, and RMB *drag* to something else. Now what would that something be?

- Moving the loop, like superpaik mentioned
- Drawing a loop, like you'd draw a selection with Ctrl+LMB

**The most desired of these two should be mapped to RMB drag**. The other one could be mapped to mid button. Some people have no mid button, but no big deal, you can still set left/right manually.

Personally I'd love the ability to quickly draw a loop. But I have a mouse, so any case is fine by me.

I also suggest we drop the "mean select" since it made redundant by having "select left" and "select right".

**Comment by lookitsnicholas** [comment]
Personally, I think that moving the loop would make a bit of sense - as the multitude of modifiers used in this area performing the same behaviour could potentially be a bit confusing; am curious to hear other opinions, however.

I agree on the drop due to redundancy - for some reason, I was under the assumption that that was already completed within this PR due to its creation.

**Comment by lookitsnicholas** [comment]
is there a way to test this release with a macOS 10.14+ package? haven't seen @LmmsBot get invoked at any time after the initial commit, and can't compile these changes locally atm

**Comment by lookitsnicholas** [comment]
Can someone rerun this build? I don't currently seem to be able to do so haha

**Comment by Spekular** [comment]
The "remapping" part of this is a half measure, so I'm closing it in favor of a later PR that won't try to future-proof.

**Pull request closed without merging**
